lib/IDETool/SyntacticMacroExpansion.cpp

```diff
@@ -0,0 +1,467 @@
+  // Compute the signature of the invocation.
+  StableHasher argHasher = StableHasher::defaultHasher();
+  for (auto arg : args)
+    argHasher.combine(StringRef(arg));
+  Fingerprint argHash(std::move(argHasher));
+
+  // Check if the current instance is usable.
+  if (auto currentInstance = this->currentInstance) {
+    if (currentInstance->getArgHash() == argHash) {
+      return currentInstance;
+    }
+  }
+
+  // Create and configure a new instance.
+  auto instance = std::make_shared<SyntacticMacroExpansionInstance>(argHash);
+
+  bool failed = instance->setup(SwiftExecutablePath, args, Plugins, error);
+  if (failed)
+    return nullptr;
+
+  currentInstance = instance;
+  return instance;
+}
+
+bool SyntacticMacroExpansionInstance::setup(
+    StringRef SwiftExecutablePath, ArrayRef<const char *> args,
+    std::shared_ptr<PluginRegistry> plugins, std::string &error) {
+  SmallString<256> driverPath(SwiftExecutablePath);
+  llvm::sys::path::remove_filename(driverPath);
+  llvm::sys::path::append(driverPath, "swiftc");
+
+  // Setup CompilerInstance to configure the plugin search path correctly.
+  bool hadError = driver::getSingleFrontendInvocationFromDriverArguments(
+      driverPath, args, Diags,
+      [&](ArrayRef<const char *> frontendArgs) {
+        return invocation.parseArgs(
+            frontendArgs, Diags, /*ConfigurationFileBuffers=*/nullptr,
+            /*workingDirectory=*/{}, SwiftExecutablePath);
+      },
+      /*ForceNoOutput=*/true);
+  if (hadError) {
+    error = "failed to setup compiler invocation";
+    return true;
+  }
+
+  // Setup filesystem.
+  auto FS = invocation.getSearchPathOptions().makeOverlayFileSystem(
+      SourceMgr.getFileSystem());
+  if (!FS) {
+    llvm::consumeError(FS.takeError());
+    error = "failed to setup overlay filesystem";
+    return true;
+  }
+  SourceMgr.setFileSystem(FS.get());
+
+  // Setup ASTContext.
+  Ctx.reset(ASTContext::get(
+      invocation.getLangOptions(), invocation.getTypeCheckerOptions(),
+      invocation.getSILOptions(), invocation.getSearchPathOptions(),
+      invocation.getClangImporterOptions(), invocation.getSymbolGraphOptions(),
+      SourceMgr, Diags));
+  registerParseRequestFunctions(Ctx->evaluator);
+  registerTypeCheckerRequestFunctions(Ctx->evaluator);
+
+  std::unique_ptr<PluginLoader> pluginLoader =
+      std::make_unique<PluginLoader>(*Ctx, /*DepTracker=*/nullptr);
+  pluginLoader->setRegistry(plugins.get());
+  Ctx->setPluginLoader(std::move(pluginLoader));
+
+  // Create a module where SourceFiles reside.
+  Identifier ID = Ctx->getIdentifier(invocation.getModuleName());
+  TheModule = ModuleDecl::create(ID, *Ctx);
+
+  return false;
+}
+
+SourceFile *
+SyntacticMacroExpansionInstance::getSourceFile(llvm::MemoryBuffer *inputBuf) {
+
+  // If there is a SourceFile with the same name and the content, use it.
+  // Note that this finds the generated source file that was created in the
+  // previous expansion requests.
```

Comment:
IIUC this will only work up until another request is received that has different arguments. Maybe that's okay, but it does mean that all requests have to be sequential for now, ie. you couldn't request for expansions in different files concurrently.

Reply:
Different _files_ still usually share the same compiler arguments as long as they are in the same module.
But yes, if compiler arguments are different, concurrent requests for nested expansion may not work.

Reply:
Do we care about any arguments other than the plugin search ones? Could we just filter to that?

Reply:
We currently use `-module-name` too as far as I remember. I don't think it's a good idea to "sanitize" compiler arguments here. Affecting options might change overtime, but I'm skeptical about anyone changing the compiler options or CompilerInvocation correctly updates this request.

Reply:
This is less sanitizing and more picking the ones we care about.

> anyone changing the compiler options or CompilerInvocation correctly updates this request

Unless it impacts it, they shouldn't need to.

Reply:
_They_ don't know if their change affects this. Like, a new plugin search option might be added in the future.

Reply:
I removed the caching logic from this PR for now.